### PR TITLE
Add fault tolerance support for trial failure

### PR DESCRIPTION
**Pull request by DeeperMind**
In practice, even the training container is correctly implemented, trials may fail due to a variety of reasons such as insufficient GPU memories. If trials fail, `GetEvaluationResult()` will return None. The suggestion should be able to handle that exception. My solution is: 

- If not all the trials failed, use the metrics of the successful ones to do the update
- If all the trials failed:
  1. Firstly, try to respawn the previous trials after sleeping for `RESPAWN_SLEEP` seconds.
  2. If respawning the trials for `RESPAWN_LIMIT` times still cannot collect valid results, then fail the task because it may indicate that the training container has errors.

Besides, this PR also fixes some small bugs and adds some important TODOs.

An example of fault handling output
```
---------------------------------------------------------------------------
Suggestion Step 1 for StudyJob nas-example-1 (ID: e9850c4a885acb19)
---------------------------------------------------------------------------
>>> 2 Trials succeeded, 1 Trials failed:
p55d6ef3720fec95: Failed
n8f94938e123b38d: 0.6475
g30851e3c259d07a: 0.6709
The average is 0.6592

>>> Suggestion updated. LSTM Controller Reward: 44.43968200683594
```

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/katib/424)

### CODE REVIEW

**Comment by DeeperMind** [comment]
\assign @hougangliu

File: pkg/suggestion/nasrl_service.py [hunk 1]
```diff
@@ -272,13 +278,36 @@ def GetSuggestions(self, request, context):
...
 
-                    # In some rare cases, GetEvaluationResult() may return None
-                    # if GetSuggestions() is called before all the trials are completed
+
+                    # Sometimes training container may fail and GetEvaluationResult() will return None
+                    # In this case, the Suggestion will:
+                    # 1. Try to call GetEvaluationResult() again
+                    # 2. If calling GetEvaluationResult() for RECALL_LIMIT times all return None, 
+                    #    then respawn the previous trials
+                    # 3. If respawning the trials for RESPAWAN_LIMIT times still cannot collect valid results,
```
Comment (andreyvelich) [inline pkg/suggestion/nasrl_service.py]: RESPAWAN_LIMIT -> RESPAWN_LIMIT

**Comment by andreyvelich** [comment]
@hougangliu @YujiOshima @johnugeorge 
Do we have any way to make StudyJob Failed from Suggestion ?

**Comment by DeeperMind** [comment]
@hougangliu This PR is well tested and ready to merge. Please take a look.

**Comment by DeeperMind** [comment]
/ assign @richardsliu @YujiOshima

**Comment by DeeperMind** [comment]
Could you approve it so that I can go on @hougangliu

**Comment by hougangliu** [comment]
/approve
thanks!

**Pull request merged**